parser/parser.go

```diff
@@ -0,0 +1,252 @@
+		log.Panicf("could not find A+ column")
+	}
+
+	for i := 1; i < len(records); i++ {
+		// convert grade distribution from string to int
+		var tempSlice bson.A
+		for j := 0; j < 13; j++ {
+			var tempInt int
+			fmt.Sscan(records[i][aPlusCol+j], &tempInt)
+			tempSlice = append(tempSlice, tempInt)
+		}
+		// add w number to the grade_distribution slice
+		var tempInt int
+		if wCol != -1 {
+			fmt.Sscan(records[i][wCol], &tempInt)
+		}
+		tempSlice = append(tempSlice, tempInt)
+		// add new class to classes slice
+		classes = append(classes,
+			model.Class{
+				Subject:           records[i][subjectCol],
+				CatalogNumber:     records[i][catalogNumberCol],
+				Section:           records[i][sectionCol],
+				GradeDistribution: tempSlice,
+			})
+	}
+	return classes
+}
+
+// inserts grades into mongodb
+func insertGrades(sectionsCollection *mongo.Collection, coursesCollection *mongo.Collection, classes []model.Class, academicSession string, logFile *os.File) {
+	for i := 0; i < len(classes); i++ {
+		var courseSearchResult model.Course
+		err := coursesCollection.FindOne(context.TODO(), bson.D{{"course_number", classes[i].CatalogNumber}, {"subject_prefix", classes[i].Subject}}).Decode(&courseSearchResult)
+		// if class is not in courses section
+		if err != nil {
+			// log that class could not be found
+			logFile.WriteString("could not find course " + classes[i].Subject + " " + classes[i].CatalogNumber + ": " + err.Error() + "\n")
+			fmt.Println("could not find course " + classes[i].Subject + " " + classes[i].CatalogNumber + ": " + err.Error())
+			// fmt.Println(err)
+			continue
+		}
+		var data []bson.M
+		match :=
+			bson.D{
+				{"$match",
+					bson.D{
+						{"course_number", classes[i].CatalogNumber},
+						{"subject_prefix", classes[i].Subject},
+					},
+				},
+			}
+		lookup :=
+			bson.D{
+				{"$lookup",
+					bson.D{
+						{"from", "sections"},
+						{"localField", "sections"},
+						{"foreignField", "_id"},
+						{"as", "sections"},
+					},
+				},
+			}
+		unwind := bson.D{{"$unwind", bson.D{{"path", "$sections"}}}}
+		matchSectionNo :=
+			bson.D{
+				{"$match",
+					bson.D{
+						{"sections.section_number", classes[i].Section},
+						{"sections.academic_session.name", academicSession},
+					},
+				},
+			}
+		project := bson.D{{"$project", bson.D{{"section", "$sections"}}}}
+		set :=
+			bson.D{
+				{"$set",
+					bson.D{
+						{"section.grade_distribution", classes[i].GradeDistribution},
+					},
+				},
+			}
+		cursor, err := coursesCollection.Aggregate(context.TODO(), mongo.Pipeline{match, lookup, unwind, matchSectionNo, project, set})
+		if err != nil {
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error() + "\n")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error())
+			continue
+		}
+		err = cursor.All(context.TODO(), &data) // put cursor data into []primitive.M
```

Comment:
Paging @AdamMcAdamson here - is the code beyond this point necessary? It seems that the parser is setting the grade distribution from the mongo pipeline. I ask this because when I run this on my local database, I get a bunch of 'recieved results 0 from aggregation, expected 1'. This makes sense to me because I don't see how a 'set' pipeline should return anything right?

Reply:
My bad, this error was happening because I had a bad .env file